include/swift/Driver/Job.h

```diff
@@ -34,48 +36,140 @@ namespace driver {
+  StringRef Base;
+
+  /// A filename that _will be passed_ to the command as a designated primary
+  /// input. Typically either equal to BaseInput or a temporary with a name
+  /// derived from the BaseInput it is related to. Also used as a key into
+  /// the DerivedOutputFileMap.
+  StringRef Primary;
+};
+
 class CommandOutput {
-  types::ID PrimaryOutputType;
-  
-  /// The primary output files of the command.
-  /// Usually a command has only a single output file. Only the compiler in
-  /// multi-threaded compilation produces multiple output files.
-  SmallVector<std::string, 1> PrimaryOutputFilenames;
 
-  /// For each primary output file there is a base input. This is the input file
-  /// from which the output file is derived.
-  SmallVector<StringRef, 1> BaseInputs;
+  /// A CommandOutput designates one type of output as primary, though there
+  /// may be multiple outputs of that type.
+  types::ID PrimaryOutputType;
 
-  llvm::SmallDenseMap<types::ID, std::string, 4> AdditionalOutputsMap;
+  /// A CommandOutput also restricts its attention regarding additional-outputs
+  /// to a subset of the PrimaryOutputs associated with its PrimaryInputs;
+  /// sometimes multiple commands operate on the same PrimaryInput, in different
+  /// phases (eg. autolink-extract and link both operate on the same .o file),
+  /// so Jobs cannot _just_ rely on the presence of a primary output in the
+  /// DerivedOutputFileMap.
+  llvm::SmallSet<types::ID, 4> AdditionalOutputTypes;
+
+  /// The set of input filenames for this \c CommandOutput; combined with \c
+  /// DerivedOutputMap, specifies a set of output filenames (of which one -- the
+  /// one of type \c PrimaryOutputType) is the primary output filename.
+  SmallVector<CommandInputPair, 1> Inputs;
+
+  /// All CommandOutputs in a Compilation share the same \c
+  /// DerivedOutputMap. This is computed both from any user-provided input file
+  /// map, and any inference steps.
+  OutputFileMap &DerivedOutputMap;
+
+  // If there is an entry in the DerivedOutputMap for a given (\p
+  // PrimaryInputFile, \p Type) pair, return a nonempty StringRef, otherwise
+  // return an empty StringRef.
+  StringRef
+  getOutputForInputAndType(StringRef PrimaryInputFile, types::ID Type) const;
+
+  /// Add an entry to the \c DerivedOutputMap if it doesn't exist. If an entry
+  /// already exists for \p PrimaryInputFile of type \p type, assert that it has
+  /// the same value as \p OutputFile.
+  void addEntryOrCheckConflict(StringRef PrimaryInputFile,
+                               types::ID type,
+                               StringRef OutputFile);
 
 public:
-  CommandOutput(types::ID PrimaryOutputType)
-      : PrimaryOutputType(PrimaryOutputType) { }
+  CommandOutput(types::ID PrimaryOutputType, OutputFileMap &Derived);
 
-  types::ID getPrimaryOutputType() const { return PrimaryOutputType; }
+  /// Return the primary output type for this CommandOutput.
+  types::ID getPrimaryOutputType() const;
```

Comment:
Why move this out of line?

Reply:
The header was getting full enough with docs, and I found myself flipping back and forth since the methods were somewhat unevenly split between inline and out; figured nothing in this is perf-sensitive and it's easier to read with everything out-of-line.